### PR TITLE
Fixes #839: web_url host not recognized in URL parsing (github.netflix.net vs git.netflix.net)

### LINKED ISSUE #839: web_url host not recognized in URL parsing (github.netflix.net vs git.netflix.net)
## Description

When a GHE host has a `web_url` that differs from the API/git `host` (e.g. Netflix: API host = `git.netflix.net`, web UI = `github.netflix.net`), issue URLs using the web UI hostname are rejected by `gru do`.

```
% gru do https://github.netflix.net/corp/ste-studio-gateway/issues/2612
[ERROR gru] Invalid issue format. Expected: <number> or <github-url>
```

### Root cause

`HostRegistry::all_hosts()` (`src/config.rs`) returns only the `host` field from each `GhHostConfig` entry — it does not include the `web_url` hostname. When `parse_github_url` and `parse_issue_info` (`src/url_utils.rs`) check whether a URL belongs to a known host, `github.netflix.net` is absent from the list even when the config has:

```toml
[github_hosts.netflix]
host = "git.netflix.net"
web_url = "https://github.netflix.net"
```

### Expected behavior

`gru do https://github.netflix.net/corp/repo/issues/42` should be accepted and treated equivalently to `gru do https://git.netflix.net/corp/repo/issues/42`, with the web URL hostname mapped back to the configured API host for all subsequent `gh` CLI calls.

## Acceptance Criteria

- [ ] `HostRegistry` (or `parse_github_url`) resolves web UI hostnames to their configured API host
- [ ] `gru do https://<web_url_host>/owner/repo/issues/N` is accepted and works end-to-end
- [ ] `is_github_url` and `parse_github_remote` in `src/git.rs` also handle web URL hostnames
- [ ] Existing tests for `HostRegistry` and `parse_github_url` pass; new tests cover the web_url→api_host mapping

### PR DESCRIPTION
**Pull request by fotoetienne**
## Summary
- Teach `HostRegistry` about `web_url` hosts: `all_url_hosts()` returns API + web UI hosts for URL matching, and `canonical_host()` maps either form back to the API host.
- URL parsing helpers (`parse_github_url`, `parse_issue_info`, `parse_pr_info`, `is_github_url`, `parse_github_remote`, `get_github_remote`) now take `&HostRegistry` and always return the canonical API host, so subsequent `gh` CLI calls target the right endpoint even when the user pasted a web UI URL like `https://github.netflix.net/corp/repo/issues/42`.
- Updated all call sites (fix/resolve, review, rebase, chat, prompt, init, resume, minion_resolver, worktree_scanner) to pass the registry directly instead of the pre-derived host slice.

## Test plan
- `cargo test` — 1205 passing, 10 ignored (pre-existing).
- `cargo clippy --all-targets -- -D warnings` — clean.
- `cargo fmt --all -- --check` — clean.
- New unit tests:
  - `config::tests::test_canonical_host_resolves_web_url_to_api_host`
  - `config::tests::test_canonical_host_without_web_url`
  - `config::tests::test_all_url_hosts_without_web_url`
  - `config::tests::test_all_url_hosts_web_url_with_port_and_path`
  - `url_utils::tests::test_parse_github_url_web_url_issue_maps_to_api_host`
  - `url_utils::tests::test_parse_github_url_web_url_pull_maps_to_api_host`
  - `url_utils::tests::test_parse_github_url_api_host_still_works_with_web_url_config`
  - `url_utils::tests::test_parse_issue_info_web_url_returns_api_host`
  - `url_utils::tests::test_parse_pr_info_rejects_issue_url_on_web_url_host`
  - `git::tests::test_is_github_url_recognizes_web_url_host`
  - `git::tests::test_parse_github_remote_web_url_maps_to_api_host`
  - `git::tests::test_parse_github_remote_api_host_still_works_with_web_url_config`
  - `git::tests::test_parse_github_remote_ssh_web_url_maps_to_api_host`

## Notes
- The previously-public `HostRegistry::all_hosts()` (API hosts only) was removed — no caller needed it after the refactor, and `all_url_hosts()` is the semantically meaningful one now. If a caller ever needs API-only hosts, they can filter `all_url_hosts()` through `canonical_host()` or we can reintroduce the method.
- While touching `parse_github_remote`, consolidated the three scheme prefixes behind a single `strip_prefix` chain. This also fixes a latent bug where `trim_start_matches` would repeatedly strip the prefix (non-observable at the time, but no longer a foot-gun).
- The `web_url` string parser accepts both `https://` and `http://` for resilience against slightly malformed configs; the normal config validation path only documents `https://`.

Fixes #839

<sub>🤖 M1ge</sub>

### CODE REVIEW

**Review by fotoetienne — COMMENTED** [review]
Reviewed the diff against issue #839 acceptance criteria and walked through the URL parsing flow.

## Summary

Clean, well-structured fix. The switch from `&[String]` to `&HostRegistry` is the right call — it moves the "which hosts map to which canonical host" knowledge into the registry rather than reconstructing it at every call site. Tests cover the main web_url → API mapping paths (issue URL, pull URL, SSH remote, HTTPS remote, wrong-resource-type rejection), plus the port/path hostname extraction edge case.

## Findings

### Nit: inconsistent error-handling for an impossible case

Both `parse_github_url` and `parse_github_remote` iterate over `all_url_hosts()` then call `canonical_host(&host)`. Since `host` came from the same registry, the lookup cannot fail — but the two sites handle it differently:

- `src/git.rs:168` uses `.expect("host from all_url_hosts is always resolvable")` — signals "unreachable".
- `src/url_utils.rs:83` uses `?`, which silently returns `None` from the parser on the same impossible branch.

Using `.expect()` (or `unwrap_or_else(|| host.clone())`) in both places would make the invariant explicit. Minor.

### Nit: `all_url_hosts()` allocates per call inside hot loops

`is_github_url` and `parse_github_remote` each call `host_registry.all_url_hosts()`, which allocates a new `Vec<String>` every invocation. For `n` remotes in `get_github_remote`, this is O(n) allocations. Not a real perf issue at current registry sizes, but since the method returns `Vec<String>` rather than an iterator, something to keep in mind if this ever becomes hot.

### Looks good

- `web_url_to_host` correctly strips scheme, port, and path. The `find(['/', ':'])` approach handles all three example shapes in the docstring and the test covers `:443/` — the tricky case where both delimiters appear.
- Accepting both `https://` and `http://` for `web_url` values is defensive without changing the documented/validated config surface — good trade-off.
- `parse_github_remote` refactor (consolidating the three scheme branches behind a single `strip_prefix` chain) is a nice cleanup and the "repeated trim" latent bug fix is worth calling out in the PR description as you did.
- Deduping in `all_url_hosts` via `!result.iter().any(|h| h == &host)` handles the edge case where a `web_url` host happens to equal an API host.
- Removing `all_hosts()` is fine — no external crates, and the refactor migrated every call site.
- Acceptance criteria from #839 are all met: web URL resolves, `is_github_url`/`parse_github_remote` both handle it, and the canonical host flows back out so downstream `gh` calls hit the API endpoint.

Overall LGTM. The nits are stylistic and don't block.

<sub>🤖 M1gi</sub>

**Comment by fotoetienne** [comment]
M1gi, addressed both nits in 154f70b:

- **Consistent error handling**: `parse_github_url` now uses `.expect("host from all_url_hosts is always resolvable")` to match `parse_github_remote`, making the invariant explicit at both sites.
- **Hoisted allocation**: `get_github_remote` computes `all_url_hosts()` once before the per-remote loop, via a new private `url_matches_any_host(url, &[String])` helper that `is_github_url` also delegates to. No behavior change, just avoids the O(n) allocations when scanning remotes.

Thanks for the careful review.

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
## 🚨 CI Fix Escalation

Automated CI fix failed after **2/2** attempts. Human intervention required.

### Failed Checks

- **Check** — check (failure)
```
Check	fail	44s	https://github.com/fotoetienne/gru/actions/runs/24584698059/job/71891193933	
Security Audit	pass	24s	https://github.com/fotoetienne/gru/actions/runs/24584698059/job/71891193911	

```

**Labels:** `gru:blocked`

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
M1gi, thanks for the review!

Both nits addressed:

- **Inconsistent error handling:** `parse_github_url` now uses `.expect("host from all_url_hosts is always resolvable")`, matching `parse_github_remote`. The invariant is now explicit at both sites instead of silently returning `None` on an impossible branch.
- **Allocation inside hot loops:** Factored the prefix check into a `url_matches_any_host(url, &[String])` helper. `get_github_remote` now computes `all_url_hosts()` once and reuses the slice across every remote line. `parse_github_remote` was also doing two allocations (one via the old `is_github_url`, one via its own loop); it now computes the list once and uses it for both the membership check and the per-host strip. `is_github_url` was dropped since it had no remaining callers.

1205 tests pass, clippy/fmt clean. Pushed as a0656c6 (on top of 154f70b which addressed the first nit).

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], thanks for the review summary. The one actionable item (the inline comment on `web_url_to_host` silently accepting malformed values) is addressed in f60e8c2: `validate_web_url` now runs at config-load time via a shared `validate_github_hosts()` helper, which is invoked from both `validate()` (used by `gru lab`) and `load_partial()` (used by `gru do` and other non-daemon commands), since those `github_hosts.*` entries drive URL parsing for every command.

<sub>🤖 M1ge</sub>

File: src/config.rs [hunk 1]
```diff
@@ -375,9 +375,65 @@ impl HostRegistry {
...
+    let rest = s
+        .strip_prefix("https://")
+        .or_else(|| s.strip_prefix("http://"))?;
+    let end = rest.find(['/', ':']).unwrap_or(rest.len());
+    let host = &rest[..end];
+    if host.is_empty() {
+        None
+    } else {
+        Some(host.to_string())
     }
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/config.rs:465]: Copilot, good catch — agreed that silent degradation is the wrong default here. Fixed in f60e8c2:

- New `validate_web_url()` helper rejects empty values, missing scheme, empty hostname, and bare hostnames (no dot), with an error message that names the offending `[github_hosts.<name>]` entry.
- The `github_hosts.*` checks are factored into `validate_github_hosts()`, and it runs on both `validate()` (for `gru lab`) and `load_partial()` (for `gru do` and other non-daemon commands) — so a misconfigured `web_url` fails fast everywhere that URL parsing runs, not just in the daemon.
- Added tests for each failure mode and a `load_partial`-path test confirming `gru do` surfaces the error too.

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], thanks for the review. The only actionable point was the inline comment about silent `web_url` misconfiguration; I've addressed it by validating `GhHostConfig.web_url` at config-load time.

`validate_github_hosts()` now runs from both `LabConfig::load()` (for `gru lab`) and `LabConfig::load_partial()` (for `gru do` and other non-daemon commands), and rejects empty values, missing scheme, empty hostname, and hostnames without a dot — each with a specific error message naming the offending entry. Tests cover each rejection path plus the happy path. Details on the inline thread.

<sub>🤖 M1ge</sub>

File: src/config.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/config.rs:465]: Copilot, good catch — a malformed `web_url` would silently be dropped by `web_url_to_host()`, making URL parsing reject web UI URLs without any explanation.

Addressed in f60e8c2 by validating `web_url` at config-load time:
- Factored the existing github_hosts checks into `LabConfig::validate_github_hosts()` and wired it into both `load()` and `load_partial()` so URL parsing failures surface regardless of whether the caller is `gru lab` or a non-daemon command like `gru do`.
- New `validate_web_url()` rejects empty values, missing scheme, empty hostname, and hostnames without a dot — each error message names the offending `[github_hosts.<name>]` entry.
- Tests cover each rejection path plus the happy path (`test_validate_github_host_rejects_web_url_missing_scheme`, `test_validate_github_host_rejects_empty_web_url`, `test_validate_github_host_rejects_web_url_empty_hostname`, `test_validate_github_host_rejects_web_url_no_dot`, `test_validate_github_host_accepts_valid_web_url`, `test_load_partial_validates_web_url`).

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
## 🚨 CI Fix Escalation

Automated CI fix failed after **2/2** attempts. Human intervention required.

### Failed Checks

- **Check** — check (failure)
```
Check	fail	47s	https://github.com/fotoetienne/gru/actions/runs/24584905328/job/71891913965	
Security Audit	pass	29s	https://github.com/fotoetienne/gru/actions/runs/24584905328/job/71891913969	

```

**Labels:** `gru:blocked`

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], addressed the one inline comment in 19edb48 — `validate_github_hosts()` now rejects `web_url` hostname collisions (two entries sharing a web_url host, or a web_url host matching another entry's API host) so `canonical_host()` is never ambiguous. Details on the inline thread.

<sub>🤖 M1ge</sub>

File: src/config.rs [hunk 2]
```diff
@@ -657,6 +722,42 @@ impl LabConfig {
...
+                    name,
+                    gh_host.host,
+                    existing_name
+                );
+            }
+            seen_hosts.insert(&gh_host.host, name);
+
+            if let Some(web_url) = &gh_host.web_url {
+                validate_web_url(name, web_url)?;
+            }
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/config.rs:800]: Copilot, good catch — two entries whose derived `web_url` hostnames collided would make `canonical_host()` depend on `HashMap` iteration order.

Addressed in 19edb48 by extending `validate_github_hosts()`:
- Entries are now iterated in sorted order so validation errors are deterministic regardless of `HashMap` order.
- Pass 1 keeps the existing per-entry checks (non-empty host, dot-containing hostname, unique API host, well-formed `web_url`).
- Pass 2 walks entries with `web_url` and rejects:
  - a `web_url` hostname that collides with another entry's API `host` (`'web_url' hostname 'X' collides with the 'host' of [github_hosts.Y]`), and
  - a `web_url` hostname that is already used by another entry's `web_url` (`'web_url' hostname 'X' is also used by [github_hosts.Y]`).
- Allows the trivial case where `web_url` hostname equals the entry's own API `host` — redundant but unambiguous.

Tests: `test_validate_github_host_rejects_duplicate_web_url_hosts`, `test_validate_github_host_rejects_web_url_colliding_with_other_api_host`, `test_validate_github_host_allows_web_url_equal_to_own_api_host`.

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], thanks for the follow-up. Addressed the inline comment about `web_url` hostname collisions — `validate_github_hosts()` now rejects any cross-entry collision (two entries with the same `web_url` hostname, or a `web_url` hostname equal to a different entry's API host), while keeping the benign case where `web_url` matches the same entry's own API host. Details on the inline thread.

<sub>🤖 M1ge</sub>

File: src/config.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/config.rs:800]: Copilot, good catch — `canonical_host()` iterates the `github_hosts` `HashMap` directly, so overlapping `web_url` hostnames would resolve nondeterministically.

Fixed in 19edb48. `validate_github_hosts()` now runs in two passes:

1. Validate each API host and build a `host -> name` index (using keys sorted for stable error ordering across runs).
2. For each `web_url`, derive the hostname and reject:
   - collisions with another entry's API host (`...web_url hostname X collides with the host of [github_hosts.other]`),
   - collisions with another entry's `web_url` hostname (`...web_url hostname X is also used by [github_hosts.other]`).

The trivial case where a `web_url` hostname equals the *same* entry's own API host is allowed, since it introduces no ambiguity.

Tests: `test_validate_github_host_rejects_duplicate_web_url_hosts`, `test_validate_github_host_rejects_web_url_colliding_with_other_api_host`, and `test_validate_github_host_allows_web_url_equal_to_own_api_host`.

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
M1ge,

Please address these clippy findings:

```
error: this `if` can be collapsed into the outer `match`
   --> src/merge_readiness.rs:464:17
    |
464 | /                 if review.user.login == pr_author {
465 | |                     author_commented = true;
466 | |                 }
    | |_________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
    = note: `-D clippy::collapsible-match` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::collapsible_match)]`
help: collapse nested if block
    |
463 ~             "COMMENTED"
464 ~                 if review.user.login == pr_author => {
465 |                     author_commented = true;
466 ~                 }
    |

error: this `if` can be collapsed into the outer `match`
   --> src/minion_registry.rs:747:17
    |
747 | /                 if actual != recorded {
748 | |                     log::debug!(
749 | |                         "PID {} recycled: recorded start_time={}, actual={}",
750 | |                         pid,
...   |
754 | |                     return false;
755 | |                 }
    | |_________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
    |
746 ~             Some(actual)
747 ~                 if actual != recorded => {
748 |                     log::debug!(
...
754 |                     return false;
755 ~                 }
```

-Stephen

**Comment by fotoetienne** [comment]
Stephen Spalding, fixed both `clippy::collapsible_match` warnings in 37d0526:
- `src/merge_readiness.rs`: folded the `pr_author` check into the `"COMMENTED"` match arm as a guard.
- `src/minion_registry.rs`: folded the `actual != recorded` check into the `Some(actual)` arm as a guard, with a `Some(_)` no-op arm to keep the PID-recycled vs same-process paths distinct.

No behavior change. All 1214 tests still pass, `cargo fmt --check` clean, `cargo clippy -- -D warnings` (with `collapsible_match` enabled) clean.

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
Stephen Spalding, fixed in 37d0526:

- `src/merge_readiness.rs:463` — folded the inner `if review.user.login == pr_author` into a guarded match arm (`"COMMENTED" if review.user.login == pr_author`). The non-matching case falls through to the existing `_` arm, which was already a no-op for the other non-author COMMENTED paths.
- `src/minion_registry.rs:744` — folded the inner `if actual != recorded` into `Some(actual) if actual != recorded`, and added an explicit `Some(_)` arm (PID still belongs to the same process) to keep the match exhaustive alongside the existing `None` arm.

`cargo clippy --all-targets -- -D warnings` and `cargo test` both clean (1214 passing, 10 ignored).

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], thanks for the follow-up review. Both inline comments addressed in 4606f41:

1. `try_load_config()` no longer swallows `load_partial()` errors — it now distinguishes "file missing" (silent, returns `None`) from "file exists but invalid" (logs a `warn!` with the file path and the validation error, still returns `None` so the command can fall back to the default registry). Tests cover missing-file, invalid-file, and registry-fallback paths.
2. `parse_github_remote`'s doc now lists all three supported schemes (HTTPS, HTTP, SSH) to match the implementation and `url_matches_any_host`.

Details on each inline thread.

<sub>🤖 M1ge</sub>

File: src/config.rs [hunk 3]
```diff
@@ -554,13 +637,17 @@ impl LabConfig {
...
+    /// parsing used by every command.
     pub(crate) fn load_partial(path: &Path) -> Result<Self> {
         let contents = fs::read_to_string(path)
             .with_context(|| format!("Failed to read config file: {}", path.display()))?;
 
         let config: LabConfig = toml::from_str(&contents)
             .with_context(|| format!("Failed to parse config file: {}", path.display()))?;
 
+        config.validate_github_hosts()?;
+
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/config.rs:678]: Copilot, fair point — swallowing the validation error with `.ok()` defeated the fail-fast work. Fixed in 4606f41.

`try_load_config()` now splits the two cases via a new `load_or_warn()` helper:

- **File doesn't exist** → returns `None` silently (normal for users who haven't run `gru init`).
- **File exists but fails to parse/validate** → logs `log::warn!("Failed to load config file {path}: {err}. Falling back to defaults; configured [github_hosts.*] entries will be ignored until this is fixed.")` and returns `None`.

I kept the `Option`/default-fallback contract rather than changing the return to `Result` — hard-failing every `gru` command on a config error felt too aggressive for a user who's just using github.com and happens to have a stale section — but the user now sees *why* their `[github_hosts.*]` entries stopped working instead of getting an opaque "Invalid issue format" later.

Tests: `test_try_load_config_missing_file_returns_none`, `test_try_load_config_invalid_file_returns_none_and_logs`, `test_load_host_registry_falls_back_when_config_invalid`.

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], thanks — both inline comments are addressed in 4606f41:

- `try_load_config()` / `load_host_registry()` no longer swallow `load_partial()` errors via `.ok()`. A new `load_or_warn()` helper distinguishes "file absent" (return `None` silently — normal pre-init state) from "file present but invalid" (log the full validation error, then return `None` so commands that don't need GHE hosts still run). Tests cover missing-file, invalid-file, and the fallback registry behavior.
- `parse_github_remote`'s docstring now lists `http://` alongside `https://` and SSH, matching what `url_matches_any_host` actually accepts.

<sub>🤖 M1ge</sub>

File: src/git.rs [hunk 4]
```diff
@@ -199,45 +206,43 @@ fn is_github_url(url: &str, github_hosts: &[String]) -> bool {
 /// - `https://<host>/owner/repo.git`
 /// - `git@<host>:owner/repo.git`
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/git.rs:275]: Copilot, good catch — the doc was stale. Updated in 4606f41 to list all three supported schemes:

```rust
/// Accepts HTTPS, HTTP, and SSH formats for any configured GitHub host:
/// - `https://<host>/owner/repo.git`
/// - `http://<host>/owner/repo.git`
/// - `git@<host>:owner/repo.git`
```

Now consistent with what `url_matches_any_host` actually checks.

<sub>🤖 M1ge</sub>

File: src/config.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/config.rs:678]: Copilot, good catch — the `.ok()` on `load_partial()` was turning "config file invalid" into "use defaults" without any sign to the user, which exactly undoes the fail-fast intent.

Fixed in 4606f41 by extracting a `load_or_warn()` helper that splits the two cases:
- File doesn't exist → return `None` silently (the normal pre-init state).
- File exists but `load_partial()` fails → `log::warn!` with the full anyhow chain (`{:#}`), then return `None` so commands that don't actually need GHE hosts still run. The warning names the path and says configured `[github_hosts.*]` entries will be ignored until it's fixed.

`try_load_config()` now goes through `load_or_warn()` for both the test-override path and the default path, and `load_host_registry()` inherits the behavior.

New tests:
- `test_try_load_config_missing_file_returns_none` — absent file stays silent.
- `test_try_load_config_invalid_file_returns_none_and_logs` — invalid file returns `None` (we can't easily capture the log output in unit tests, but we assert invalid config doesn't masquerade as valid).
- `test_load_host_registry_falls_back_when_config_invalid` — registry drops the malformed entry instead of silently accepting it.

<sub>🤖 M1ge</sub>

File: src/git.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/git.rs:275]: Copilot, correct — `url_matches_any_host` has been checking `http://` alongside `https://` and `git@…:` all along, the docstring just hadn't caught up.

Updated in 4606f41: the doc comment now reads "Accepts HTTPS, HTTP, and SSH formats" and the example list includes `http://<host>/owner/repo.git` so readers see all three schemes the implementation actually handles.

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], thanks for the follow-up. Addressed the inline comment in 5eec678 — the prefix-matching approach in `url_matches_any_host` and `parse_github_url` was the same underlying issue you raised at the top level: any `:port` in the authority meant the URL was treated as non-GitHub, and `parse_github_url` was stricter about `http://` than the config validation was.

Replaced the string prefix approach with a `split_github_url()` helper that parses a URL into (scheme, hostname, rest), stripping an optional `:port` from the authority. `url_matches_any_host`, `parse_github_remote`, and `parse_github_url` now all use it. `parse_github_url` also now accepts `http://` to align with `validate_web_url`. Details on the inline thread.

<sub>🤖 M1ge</sub>

File: src/url_utils.rs [hunk 5]
```diff
@@ -33,19 +34,24 @@ fn clean_url(url: &str) -> &str {
...
 ///
 /// Returns `None` if the URL is not a valid GitHub issue/PR URL.
-pub(crate) fn parse_github_url(url: &str, github_hosts: &[String]) -> Option<GitHubUrl> {
+pub(crate) fn parse_github_url(url: &str, host_registry: &HostRegistry) -> Option<GitHubUrl> {
     let cleaned = clean_url(url);
 
-    for host in github_hosts {
+    for host in host_registry.all_url_hosts() {
         let prefix = format!("https://{}/", host);
         if let Some(path) = cleaned.strip_prefix(&prefix) {
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/url_utils.rs]: Copilot, you're right on both counts — `parse_github_url` only ever saw the URL through a `https://<host>/` prefix match, so URLs with `http://` or an explicit `:port` fell through even when the host was configured. And that diverged from `validate_web_url` / `web_url_to_host`, which have been lenient about both.

Fixed in 5eec678. Introduced a `split_github_url()` helper in `git.rs` that parses a URL into `(scheme, hostname, rest)`, stripping an optional `:port` from the authority and rejecting userinfo (`user@host`) and empty hosts. `url_matches_any_host`, `parse_github_remote`, and `parse_github_url` now all go through it.

Specifically for `parse_github_url`:
- Accepts `http://` (matches `validate_web_url`).
- Accepts `https://host:port/...` (port stripped before lookup).
- Continues to reject SSH URLs (they're git remote URLs, not web URLs).
- Existing `test_parse_github_url_rejects_http` replaced by `test_parse_github_url_accepts_http_scheme`. New `test_parse_github_url_accepts_explicit_port` and `test_parse_github_url_rejects_ssh` lock in the new contract.

On the remote side: new tests cover `https://ghe.example.com:8443/...` and `http://ghe.example.com/...` as valid remotes, plus `split_github_url` has its own battery of unit tests (port, http, ssh, bare host, userinfo, empty host, unsupported scheme).

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], thanks — both the suppressed remote-URL comment and the inline `parse_github_url` comment describe the same underlying issue (string-prefix matching missing optional `:port`, and `http://` accepted inconsistently), so I fixed them together in 5eec678:

- New `split_github_url(url) -> Option<GitUrlParts>` in `src/git.rs` parses the URL into `(scheme, host-without-port, remainder)` for HTTPS, HTTP, and SSH shapes.
- `url_matches_any_host`, `parse_github_remote`, and `parse_github_url` all flow through the helper. Ports are stripped before host comparison, so `https://ghe.example.com:8443/org/repo.git` matches when `ghe.example.com` is configured.
- `parse_github_url` now accepts `http://` alongside `https://`, matching what `validate_web_url` / `web_url_to_host` allow. SSH-style URLs remain rejected by the web URL parser since they're remote URLs.
- HTTP userinfo (`https://user@host/...`) is rejected explicitly to prevent homograph-style tricks where an attacker embeds the configured host in the userinfo portion.

Tests: `test_split_github_url_https_with_port`, `test_split_github_url_ssh`, `test_split_github_url_rejects_userinfo`, `test_split_github_url_rejects_empty_host`, `test_split_github_url_rejects_unsupported_scheme`, `test_parse_github_remote_https_with_explicit_port`, `test_parse_github_remote_http_scheme`, `test_parse_github_url_accepts_explicit_port`, `test_parse_github_url_accepts_http_scheme`. The old `test_parse_github_url_rejects_http` was flipped to reflect the new contract.

<sub>🤖 M1ge</sub>

File: src/url_utils.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/url_utils.rs]: Copilot, good catch — the asymmetry between `validate_web_url` (which accepts `http://` and ports) and `parse_github_url` (which only accepted `https://` with no port) meant perfectly valid self-hosted GHE URLs would be rejected despite having a matching `web_url` entry.

Fixed in 5eec678 by replacing the per-host prefix scanning with a shared `split_github_url()` helper in `src/git.rs` that returns `(scheme, host-without-port, remainder)`. `parse_github_url` now:

- Accepts both `http://` and `https://` schemes.
- Strips any `:port` from the authority before looking the host up in the registry, so `https://ghe.example.com:8443/org/repo/issues/1` parses correctly.
- Rejects SSH-style URLs (they're remote URLs, not web URLs).
- Rejects HTTP userinfo (`https://user@host/...`) to prevent an attacker from hiding a configured host in the userinfo portion.

Tests added: `test_parse_github_url_accepts_explicit_port`, `test_parse_github_url_accepts_http_scheme`, and supporting `split_github_url` tests covering ports, SSH, userinfo rejection, empty hosts, and unsupported schemes. The old `test_parse_github_url_rejects_http` was flipped to reflect the new contract.

<sub>🤖 M1ge</sub>

**Comment by fotoetienne** [comment]
🧑‍⚖️ **Merge readiness: 3/10 — needs human review**

PR has an active CI fix escalation — automated CI fix failed after 2/2 attempts and the PR was labeled gru:blocked. Additionally, the human owner (Stephen) posted a comment at 21:40 asking M1ge to address clippy findings (collapsible_match in merge_readiness.rs and minion_registry.rs). This is the most recent interaction and represents unaddressed reviewer feedback requiring code changes, not just a reply. Cannot merge until clippy issues are fixed and CI passes.

_To proceed, remove the `gru:needs-human-review` label. The judge will re-evaluate on the next PR state change._

**Comment by fotoetienne** [comment]
🧑‍⚖️ **Merge readiness: 3/10 — needs human review**

Two CI Fix Escalation comments report automated CI fix failed after 2/2 attempts with gru:blocked label applied. The human maintainer (Stephen) posted a comment at 21:40 requesting clippy findings be addressed (collapsible_match warnings in merge_readiness.rs and minion_registry.rs). While commit 37d0526 claims to fix clippy::collapsible_match warnings, the escalation state and unresolved maintainer request indicate this needs human verification before merge.

_To proceed, remove the `gru:needs-human-review` label. The judge will re-evaluate on the next PR state change._

**Comment by fotoetienne** [comment]
CI looks good!

**Comment by fotoetienne** [comment]
---
type: monitoring-paused
---

⏸️ This PR's automated agent has paused. There are 2 reviews that haven't been addressed yet. Resume automated responses with:
`gru resume M1ge`

<sub>🤖 M1ge</sub>
